2.4 xb 1548

altmannmarcelo opened the pull request:
This implements #PXB-1548.

Buffer pool dump happens at the beginning of backup if `--dump-innodb-buffer-pool` is set. 
Prior to copying `ib_buffer_pool` it checks if current `SHOW STATUS LIKE 'Innodb_buffer_pool_dump_status'` contains the string `dump completed at`. If dump is still in progress, the tool keeps checking untill `--dump-innodb-buffer-pool-timeout=N` seconds has elapsed since we originally started the dump.

User can choose to change the default `innodb_buffer_pool_dump_pct`. If `--dump-innodb-buffer-pool-pct` is set, it stores the current MySQL `innodb_buffer_pool_dump_pct` value, then it changes it to the desired percentage. After the check in the end of backup, orginal values is restored back.

3 new parameters have been added:
`--dump-innodb-buffer-pool` - When set, this will trigger dump of buffer pool prior to copying ib_buffer_pool
`--dump-innodb-buffer-pool-timeout=N` - This controlls for how long (in seconds) to wait for dump of buffer pool before moving on. (default: 10)
`--dump-innodb-buffer-pool-pct=N` - This controlls the percentage of buffer pool to dump. If set, xtrabackup will save current MySQL value, change to the desired value, once dump is completed or it aborts due to time out, it will change it back to original value.

altmannmarcelo committed:
This commit implements #PXB-1548

3 new parameters have been added:
--dump-innodb-buffer-pool - When set, this will trigger dump of buffer pool prior to copying ib_buffer_pool
--dump-innodb-buffer-pool-timeout=N - This controlls for how long (in seconds) to wait for dump of buffer pool before moving on. (default: 10)
--dump-innodb-buffer-pool-pct=N - This controlls the percentage of buffer pool to dump. If set, xtrabackup will save current MySQL value, change to the desired value, once dump is completed or it aborts due to time out, it will change it back to original value.

altmannmarcelo committed:
Adjusted Innodb Buffer pool dump. Now it is called
at the beginning of backup and status is checked right
before copying dump file.

altmannmarcelo committed:
Whitespace adjusts

gl-sergei commented:
Hi @altmannmarcelo ! Are you willing to continue to work on this patch? If so, I would ask to you to start with adding a test case (you can see examples in xtrabackup/test/t/ directory). Then fix code formatting. Please also make sure to check that patch doesn't break anything on older MySQL versions.

altmannmarcelo committed:
Added test cases.
Adjusted code format
Replaced sleep by os_thread_sleep
Added a function to only run the dump if
`SET GLOBAL innodb_buffer_pool_dump_now=ON` is supported

altmannmarcelo commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+			"SHOW GLOBAL VARIABLES LIKE 'innodb_buffer_pool_dump_pct'", 
+			variables, true);
+
+		original_innodb_buffer_pool_dump_pct = 
```
@gl-sergei I'm not sure is ok to break the value assignment into 2 line, but leaving it on a single line would exceed 80 characters.

gl-sergei replied:
trailing whitespace

altmannmarcelo committed:
added a check to not check from dump if it is not supported

altmannmarcelo commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+			variables, true);
+
+		original_innodb_buffer_pool_dump_pct = 
+		atoi(innodb_buffer_pool_dump_pct);
```
@gl-sergei I'm not sure is ok to break the value assignment into 2 line, but leaving it on a single line would exceed 80 characters.

gl-sergei replied:
It is fine to break, please also add extra tab on second line

altmannmarcelo commented:
Tested on not supported version and it works fine. Even if you pass `--dump-innodb-buffer-pool` it won't attempt to run the dump.

gl-sergei commented on storage/innobase/xtrabackup/src/backup_copy.cc:
```diff
@@ -1559,6 +1559,11 @@ backup_finish()
 
 	/* Copy buffer pool dump or LRU dump */
 	if (!opt_rsync) {
+		if(opt_dump_innodb_buffer_pool && has_innodb_buffer_pool_dump()) {
```
s/if(opt/if (opt/

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+bool
+has_innodb_buffer_pool_dump()
+{
+	if((server_flavor == FLAVOR_PERCONA_SERVER || 
```
s/if((server_flavor/if ((server_flavor/

gl-sergei replied:
trailing whitespace

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+		return true;
+	}
+		
+	if(server_flavor == FLAVOR_MARIADB && mysql_server_version >= 10000) {
```
s/if(server_flavor/if (server_flavor/

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+bool
+has_innodb_buffer_pool_dump_pct()
+{
+	if((server_flavor == FLAVOR_PERCONA_SERVER || 
```
s/if((server_flavor/if ((server_flavor/

gl-sergei replied:
trailing whitespace

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+		return true;
+	}
+		
+	if(server_flavor == FLAVOR_MARIADB && mysql_server_version >= 10110) {
```
s/if(server_flavor/if (server_flavor/

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+	char *change_bp_dump_pct_query;
+
+	/* Verify if we need to change innodb_buffer_pool_dump_pct */
+	if(opt_dump_innodb_buffer_pool_pct != 0 && 
```
s/if(opt_dump_innodb_buffer_pool_pct/if (opt_dump_innodb_buffer_pool_pct/

gl-sergei replied:
trailing whitespace

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+	if(opt_dump_innodb_buffer_pool_pct != 0 && 
+		has_innodb_buffer_pool_dump_pct()) {
+		mysql_variable variables[] = {
+			{"innodb_buffer_pool_dump_pct", &innodb_buffer_pool_dump_pct},
```
Line is too long

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+			{NULL, NULL}
+		};
+		read_mysql_variables(connection,
+			"SHOW GLOBAL VARIABLES LIKE 'innodb_buffer_pool_dump_pct'", 
```
Line is too long and extra trailing space

gl-sergei replied:
trailing whitespace

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+		xb_a(asprintf(&change_bp_dump_pct_query,
+				"SET GLOBAL innodb_buffer_pool_dump_pct = %u",
+				opt_dump_innodb_buffer_pool_pct));
+		xb_mysql_query(mysql_connection, change_bp_dump_pct_query, false);
```
This line is too long

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+
+
+	mysql_variable status[] = {
+			{"Innodb_buffer_pool_dump_status", &innodb_buffer_pool_dump_status},
```
This line and entire next block are too long

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+	free_mysql_variables(status);
+
+	/* restore original innodb_buffer_pool_dump_pct */
+	if(opt_dump_innodb_buffer_pool_pct != 0 && 
```
s/if(opt_dump_innodb_buffer_pool/if (opt_dump_innodb_buffer_pool/

gl-sergei replied:
trailing whitespace

gl-sergei commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -4712,6 +4737,12 @@ xtrabackup_backup_func(void)
 	}
 
+	if (!opt_rsync) {
+		if(opt_dump_innodb_buffer_pool && has_innodb_buffer_pool_dump()) {
```
s/if(opt_dump_innodb_buffer_pool/if (opt_dump_innodb_buffer_pool/

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+		server_flavor == FLAVOR_MYSQL) && mysql_server_version >= 50702) {
+		return true;
+	}
+		
```
Whitespace on empty line.

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+		server_flavor == FLAVOR_MYSQL) && mysql_server_version >= 50603) {
+		return true;
+	}
+		
```
Whitespace on empty line

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+	/* check if dump has been completed */
+	msg_ts("Checking if InnoDB buffer pool dump has completed\n");
+	while (!strstr(innodb_buffer_pool_dump_status, "dump completed at")) {
+		if(innodb_buffer_pool_dump_start_time + 
```
Missing space after "if" and extra trailing whitespace

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
+	return false;
+}
+
+bool
```
Please make it static

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.h:
```diff
@@ -113,4 +113,16 @@ mdl_lock_table(ulint space_id);
+bool
+has_innodb_buffer_pool_dump();
+
+bool
```
Please make it static and local to `backup_mysql`

gl-sergei commented on storage/innobase/xtrabackup/src/backup_mysql.h:
```diff
@@ -113,4 +113,16 @@ mdl_lock_table(ulint space_id);
 mdl_unlock_all();
 
+bool
+has_innodb_buffer_pool_dump();
```
Consider making it static and local to `backup_mysql`, `check_dump_buffer_pool` could be made the only entry point.

gl-sergei commented on storage/innobase/xtrabackup/src/backup_copy.cc:
```diff
@@ -1559,6 +1559,11 @@ backup_finish()
 
 	/* Copy buffer pool dump or LRU dump */
 	if (!opt_rsync) {
```
Can we make it work with rsync?

altmannmarcelo committed:
Adjusted formatting
Added feature to also works with rsync

altmannmarcelo commented:
@gl-sergei adjusted code as per your review

altmannmarcelo committed:
adjusted memory leak and issue with debug build

altmannmarcelo commented:
Implemented in a single commit over https://github.com/percona/percona-xtrabackup/pull/563 . Thanks, @gl-sergei for the review.
